Initial contribution of the PioneerAVR binding

Stratehm opened the pull request:
Signed-off-by: Antoine Besnard <stratehm@hotmail.com> (github:
@Stratehm)

kaikreuzer commented:
Thanks for this contribution! I hope I will soon find the time to review it, there is quite some backlog of binding PRs by now... I'll do my best!

kaikreuzer commented on addons/binding/org.openhab.binding.pioneeravr/ESH-INF/i18n/pioneeravr_xx_XX.properties:
```diff
@@ -0,0 +1,11 @@
+# binding
```
please remove this file, it is only a template if you want to provide other locales.

kaikreuzer commented on addons/binding/org.openhab.binding.pioneeravr/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,123 @@
+	<!-- AVR control over IP -->
+	<thing-type id="ipAvr">
+		<label>Pioneer AVR over IP</label>
+		<description>Control a Pioneer AVR over through a Serial port </description>
```
should be "over IP", not "over through Serial port "

kaikreuzer commented on addons/binding/org.openhab.binding.pioneeravr/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,123 @@
+	<!-- AVR control over Serial -->
+	<thing-type id="serialAvr">
+		<label>Pioneer AVR over Serial</label>
+		<description>Control a Pioneer AVR over through a Serial port (RS-232)</description>
```
remove the word "through"

kaikreuzer commented on addons/binding/org.openhab.binding.pioneeravr/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,123 @@
+		<item-type>Dimmer</item-type>
+		<label>Volume</label>
+		<description>Increase/Decrease the volume (%) and mute/un-mute</description>
+		<category>Dimmer</category>
```
category should be SoundVolume

kaikreuzer commented on addons/binding/org.openhab.binding.pioneeravr/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,123 @@
+		<label>Volume</label>
+		<description>Increase/Decrease the volume (%) and mute/un-mute</description>
+		<category>Dimmer</category>
+		<state min="0" max="100" />
```
you might also want to provide a pattern like "%d %%"

kaikreuzer commented on addons/binding/org.openhab.binding.pioneeravr/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,123 @@
+		<label>Volume</label>
+		<description>Set the volume level (dB)</description>
+		<category>SoundVolume</category>
+		<state min="-80" max="12" step="0.5" />
```
add pattern="%.1f dB"

kaikreuzer commented on addons/binding/org.openhab.binding.pioneeravr/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,123 @@
+		<item-type>String</item-type>
+		<label>Input source</label>
+		<description>Select the input source of the AVR</description>
+		<state pattern="[0-9]{2}">
```
The pattern is meant for UIs, which would display the option texts and not the keys, so probably better remove the pattern here.

kaikreuzer commented on addons/binding/org.openhab.binding.pioneeravr/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,27 @@
+Bundle-ClassPath: .
+Import-Package: com.google.common.collect,
+ gnu.io;resolution:=optional,
+ org.apache.commons.lang;version="2.6.0",
```
please remove all version constraints

kaikreuzer commented on addons/binding/org.openhab.binding.pioneeravr/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,27 @@
+Service-Component: OSGI-INF/*.xml
+Export-Package: org.openhab.binding.pioneeravr,
+ org.openhab.binding.pioneeravr.discovery,
+ org.openhab.binding.pioneeravr.handler
```
you should either also export org.openhab.binding.pioneeravr.protocol or move its classes into an internal package.

kaikreuzer commented on addons/binding/org.openhab.binding.pioneeravr/OSGI-INF/PioneerAvrHandlerFactory.xml:
```diff
@@ -0,0 +1,17 @@
+   <service>
+      <provide interface="org.eclipse.smarthome.core.thing.binding.ThingHandlerFactory"/>
+   </service>
+   <property name="service.pid" type="String" value="org.openhab.binding.pioneeravr"/>
```
do you need this service pid?

kaikreuzer commented on addons/binding/org.openhab.binding.pioneeravr/OSGI-INF/PioneerAvrDiscovery.xml:
```diff
@@ -0,0 +1,19 @@
+      <provide interface="org.eclipse.smarthome.config.discovery.UpnpDiscoveryParticipant"/>
+      <provide interface="org.osgi.service.cm.ManagedService"/>
+   </service>
+   <property name="service.pid" type="String" value="org.openhab.binding.pioneeravr"/>
```
the service pids should rather be org.openhab.xxx - but here I would suggest to remove it completely and adapt the component name instead.

kaikreuzer commented on addons/binding/org.openhab.binding.pioneeravr/OSGI-INF/PioneerAvrDiscovery.xml:
```diff
@@ -0,0 +1,19 @@
+    http://www.eclipse.org/legal/epl-v10.html
+
+-->
+<scr:component xmlns:scr="http://www.osgi.org/xmlns/scr/v1.1.0" immediate="true" name="org.openhab.binding.pioneeravr.discovery">
```
you could change the name to org.openhab.pioneeravr in order to react to this pid for the ConfigurationAdmin

kaikreuzer commented on addons/binding/org.openhab.binding.pioneeravr/OSGI-INF/PioneerAvrDiscovery.xml:
```diff
@@ -0,0 +1,19 @@
+   <implementation class="org.openhab.binding.pioneeravr.discovery.PioneerAvrDiscoveryParticipant"/>
+   <service>
+      <provide interface="org.eclipse.smarthome.config.discovery.UpnpDiscoveryParticipant"/>
+      <provide interface="org.osgi.service.cm.ManagedService"/>
```
no need to implement ManagedService if you use passing config information through the activate method.

kaikreuzer commented on addons/binding/org.openhab.binding.pioneeravr/README.md:
````diff
@@ -0,0 +1,109 @@
+The binding can auto-discover the Pioneer AVRs present on your local network. The auto-discovery is enabled by default. To disable it, you can create a file in the services directory called pioneeravr.cfg with the following content:
+
+```
+ pid:org.openhab.binding.pioneeravr
````
if you apply the above suggestions, you can remove this line

kaikreuzer commented on addons/binding/org.openhab.binding.pioneeravr/plugin.xml:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This file should be deleted.

kaikreuzer commented on addons/binding/org.openhab.binding.pioneeravr/src/main/java/org/openhab/binding/pioneeravr/discovery/PioneerAvrDiscoveryParticipant.java:
```diff
@@ -0,0 +1,103 @@
+ * @author Antoine Besnard
+ *
+ */
+public class PioneerAvrDiscoveryParticipant implements UpnpDiscoveryParticipant, ManagedService {
```
Don't use ManagedService, but rather use DS to provide configurations, see http://stackoverflow.com/questions/4037388/osgi-declarative-services-vs-managedservice-for-configuring-service

kaikreuzer commented on addons/binding/org.openhab.binding.pioneeravr/src/main/java/org/openhab/binding/pioneeravr/discovery/PioneerAvrDiscoveryParticipant.java:
```diff
@@ -0,0 +1,103 @@
+
+	@Override
+	public Set<ThingTypeUID> getSupportedThingTypeUIDs() {
+		return supportedThingTypes;
```
If no configuration is provided, this will return an empty Set, although discovery is enabled by default!

Stratehm replied:
Some people may want to disable the PioneerAvr auto discovery without disabling the whole Upnp auto discovery. To do so, the PioneerAvrDiscoveryParticipant return an empty Set of supported Thing types, thus PioneerAvrs will never be discovered (even if the Upnp discovery process actually find them). It is the only way I have found to do so. 

Another solution may be to add a function to UpnpDiscoveryParticipant to enable/disable a Participant like:
```
public boolean isEnabled();
```

kaikreuzer replied:
My point was that the default (i.e. if NO configuration is given at all) should be enabled as it is a rather unusual case to deactivate discovery (note that this is what the inbox is for: Discovery does not mean that it is added to the system, but it first needs to be approved).

Even with the current code, this is still not the case: As you do a check for ```componentContext.getProperties() != null```, the returned supportedThingTypes will be empty, although your default is ```isAutoDiscoveryEnabled = true```. You need to make sure that the supportedThingTypes are correctly filled in that case.

Stratehm replied:
Oh yes, I see. My bad. Fixed.

Thanks

kaikreuzer commented on addons/binding/org.openhab.binding.pioneeravr/src/main/java/org/openhab/binding/pioneeravr/protocol/AvrConnectionException.java:
```diff
@@ -0,0 +1,38 @@
+ * 
+ * @author Rainer Ostendorf
+ * @author based on the Onkyo binding by Pauli Anttila and others
+ * @since 1.4.0
```
please remove the since tag here.

kaikreuzer commented on addons/binding/org.openhab.binding.pioneeravr/src/main/java/org/openhab/binding/pioneeravr/protocol/event/AvrUpdateListener.java:
```diff
@@ -0,0 +1,28 @@
+ * @author Antoine Besnard
+ * @author Rainer Ostendorf
+ * @author based on the Onkyo binding by Pauli Anttila and others
+ * @since 1.4.0
```
please remove the since tag

kaikreuzer commented:
Thanks, I've finished my review.
Really very nice and clean code, respect! I only had some minor comments.
Please let me know as soon as you have incorporated them. Thanks :+1:

buildhive commented:
[openhab » openhab2 #32](https://buildhive.cloudbees.com/job/openhab/job/openhab2/32/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[openhab » openhab2 #33](https://buildhive.cloudbees.com/job/openhab/job/openhab2/33/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

Stratehm commented:
I have incorporated the changes. 
Maybe we should discuss about the auto discovery configuration of the binding. (I have added a note above)

Thanks for the review.

buildhive commented:
[openhab » openhab2 #34](https://buildhive.cloudbees.com/job/openhab/job/openhab2/34/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

Stratehm commented:
All comments have been processed. It should be ready for integration.
Thanks

buildhive commented:
[openhab » openhab2 #90](https://buildhive.cloudbees.com/job/openhab/job/openhab2/90/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[openhab » openhab2 #91](https://buildhive.cloudbees.com/job/openhab/job/openhab2/91/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[openhab » openhab2 #92](https://buildhive.cloudbees.com/job/openhab/job/openhab2/92/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

kaikreuzer commented on addons/binding/org.openhab.binding.pioneeravr/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,31 @@
+ org.slf4j
+Service-Component: OSGI-INF/*.xml
+Export-Package: org.openhab.binding.pioneeravr,
+ org.openhab.binding.pioneeravr.discovery,
```
Not sure, why I did not comment on this before, sorry: Is there any reason why  discovery and protocol packages are exported? If not, I'd suggest to rather move them to an internal package.

kaikreuzer commented on addons/binding/org.openhab.binding.pioneeravr/src/main/java/org/openhab/binding/pioneeravr/protocol/CommandTypeNotSupportedlException.java:
```diff
@@ -0,0 +1,36 @@
+ * 
+ * @author Antoine Besnard
+ */
+public class CommandTypeNotSupportedlException extends Exception {
```
This is a typo, right? It should be CommandTypeNotSupportedException not CommandTypeNotSupportedlException.

kaikreuzer commented:
Thanks, apart from the two minor comments just now, the code looks good and is ready to be merged.
Once you have done the final changes, could you please squash your commits into one? Thanks!

buildhive commented:
[openhab » openhab2 #94](https://buildhive.cloudbees.com/job/openhab/job/openhab2/94/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[openhab » openhab2 #98](https://buildhive.cloudbees.com/job/openhab/job/openhab2/98/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[openhab » openhab2 #99](https://buildhive.cloudbees.com/job/openhab/job/openhab2/99/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

Stratehm commented:
The 2 comments are fixed and all is squashed in a single commit.

dnobel commented:
Hi,

I have tested the binding with my Pioneer VSX 828 receiver. This is my feedback:

1) The declarative DS files refer non-existing classes. I guess it is because of refactoring to internal packaging. The references in the XML files must be adapted

2) My VSX-828 is found, but after I added it to my installation I see the following errors:

```
[.o.b.p.i.p.StreamAvrConnection:102  ] - Can't connect to 192.168.178.25:23. Cause: Connection refused
2015-05-08 09:59:15 [DEBUG] [.c.thing.internal.ThingManager:141  ] - Thing status of thing pioneeravr:ipAvr:5F9EC1B3-ED59-79BF-4530-745e1c301591 changed: StatusInfo [getStatus()=OFFLINE, getStatusDetail()=NONE, getDescription()=null]
```

The IP address 192.168.178 is correct. I restarted the runtime several times and also the receiver. I could not get it working. Any idea?

3) In the discovery result another property for the ip address is used (host), as in the thing type XML file (address). If you pair the device from the Inbox with the Paper UI and try to configure it afterwards, the IP Adress field is empty. Configuration values in the DiscoveryResults should match to the configuration parameters defined in the XML files.

Regards

Dennis

Stratehm commented:
Hi, 
thank you for the feedback. I will look at it in few days.

dnobel commented:
Hi

I investigated the problem and did a portscan at my VSX-828. The telnet port is not 23, but 8102.

With the correct port, I was able to get the binding working, but it looks like the protocol is a little bit different. For example, the channel "display" was not working. Moreover I was not able to change the volume by setting a concrete value (seems the device does not support that at all). If I switch my receiver to OFF, it has a state PWR02, but the binding only knows PWR01 and PWR02 and logs "Unkown response type from AVR".

With which device did you test it? Is there some kind of protocol specification?

Regards

Dennis

kaikreuzer commented:
@Stratehm, please keep in mind that this binding is scheduled for alpha 2 release on May 25; so we should try to merge a working state within the next week. So if you cannot get Dennis model supported, maybe we could start without it - but it would be good to find out, which models actually work and which don't and document this.

Stratehm commented:
@dnobel, I am using the file http://www.pioneerelectronics.com/StaticFiles/Custom%20Install/RS-232%20Codes/Av%20Receivers/Elite%20&%20Pioneer%20FY13AVR%20IP%20&%20RS-232%205-8-12.xls to have a list of commands. Unfortunately, there are only a small part of the Pioneer models described in that file.

I am testing the binding on a VSX-921 (which is not listed in the file but seems to be compatible with a large part of the commands). The VSX-828 model is not listed either.

The VSX-921 responds to telnet connections on both ports 23 and 8102. Since 23 is the official port, I choose this one as default, but maybe I should set it to 8102.

As you can see in the file, the PWR02 response is not listed and my AVR never sent it back to me. I really do not know what state it reflects (maybe a standby mode).

I have fixed the Declarative Services package names and the IP address of the Discovery result (thank you).

With my VSX-921, all channel works with PaperUI (except that the Input Source channel display the option values instead of the option names configured in the channel type)

I saw that there are 3 officials Android app to control pioneerAvr. It is not the same app for the VSX-921 and the VSX-828, so it may not be the protocol...I will try to look at it.

@kaikreuzer I will not be able to give you a compatibility list. I can test the binding only on my VSX-921. I think we will have to rely on users' expermients to build such a list.

buildhive commented:
[openhab » openhab2 #128](https://buildhive.cloudbees.com/job/openhab/job/openhab2/128/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

kaikreuzer commented:
> I will not be able to give you a compatibility list. I can test the binding only on my VSX-921. I think we will have to rely on users' expermients to build such a list.

Yes, my thinking was rather along the lines to say that we can have a dedicated thing type for the devices that we know that work and one for all others - this could be useful especially for the auto-discovery, because you might not want to pretend you that the binding supports all Pioneer AVRs. Do the AVRs send their concrete model in the UPnP information? If so, you could match this against a list of tested devices and either create a "supported" discovery result or an "unsupported" discovery result. What do you think about that?

Stratehm commented:
Good idea. The model name in the UPnP information seems to be right. I will implement this feature.

Another question, the port number to control the AVR is sent in the UPnP description file with a specific namespace. Can I get the XML tag value with a specific namespace with JUPnP ? I need to access this value: 
```
<av:X_ipRemoteTcpPort xmlns:av="http://www.pioneerelectronics.com/xmlns/av">8102</av:X_ipRemoteTcpPort>
```

Thanks

kaikreuzer commented:
I assume you have already checked the "RemoteDevice" that you receive as a UPnPDiscoveryParticipant. If the custom namespace infos are not contained in there anywhere, it might be tough to get hold of this information. You might have to register at jUPnP yourself directly or even have to implement some custom parsing; so better go for dedicated thing types for devices that are supported and where you can declare a default for the port.

Stratehm commented:
Thanks. I will look at that later.

Some modifications in the last commit:
-A new thing type is defined:  ipAvrUnsupported. All AVRs that are not in the supported list will be bound to a thing of this type. I have looked into the official Pioneer Android app and extracted a list of AVR that all use the same protocol as the VSX-921 (The dnobel's one is not in this list).

-The default port is now 8102 (it is the default one in the Android app).

-When the AVR is turned OFF, the channels' states are set to Undefined.

@dnobel the PWR2 state is a Network Standby mode. Only some AVRs send this state. I added it to the binding.

buildhive commented:
[openhab » openhab2 #129](https://buildhive.cloudbees.com/job/openhab/job/openhab2/129/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

kaikreuzer commented:
Thanks, very cool! I have just merged it. FYI: I did one minor change before merging it: I declared the dB volume channel to be advanced, so that by default you do not have two volume sliders.

kaikreuzer commented:
@Stratehm Thanks for your wonderful documentation for this binding. I have taken this as a starting point to also ask other binding developers to provide documentation:
https://github.com/eclipse/smarthome/blob/master/docs/sources/development/extensions/bindings/docs.md
As an example how the result could look like, see https://github.com/eclipse/smarthome/blob/master/addons/binding/org.eclipse.smarthome.binding.hue/README.md.

You might want to check whether you want to bring your documentation in-line with this template; but I think it is pretty good as it is.
